host/fs/fs_linux.go

```diff
@@ -114,3 +118,259 @@ func (e *event) wait(timeoutms int) (int, error) {
+	e.mu.Lock()
+	if e.epollFd == 0 {
+		// Was already closed.
+		e.mu.Unlock()
+		return nil
+	}
+	e.closing = true
+	e.mu.Unlock()
+	// Wake up the poller so it notices it should quit.
+	e.wakeUpLoop(nil)
+	// It is important to wait for the loop() function to exit, as EpollWait()
+	// will not return on a closed file handle.
+	e.wg.Wait()
+	e.mu.Lock()
+	_ = e.w.Close()
+	e.w = nil
+	_ = e.r.Close()
+	e.r = nil
+	err := syscall.Close(e.epollFd)
+	e.epollFd = 0
+	e.mu.Unlock()
+	return err
+}
+
+// loop is the main event loop.
+func (e *eventsListener) loop() {
+	defer e.wg.Done()
+	var events []syscall.EpollEvent
+	type lookup struct {
+		c     chan<- time.Time
+		event epollEvent
+	}
+	var lookups []lookup
+	for first := true; ; {
+		if !first {
+			e.mu.Lock()
+		}
+		if len(events) < len(e.fds) {
+			events = make([]syscall.EpollEvent, len(e.fds))
+		}
+		closing := e.closing
+		e.mu.Unlock()
+		first = false
+
+		if closing {
+			// We're done.
+			return
+		}
+		if len(events) == 0 {
+			panic("internal error: there's should be at least one pipe")
+		}
+
+		// http://man7.org/linux/man-pages/man2/epoll_wait.2.html
+		n, err := syscall.EpollWait(e.epollFd, events, -1)
+		if n <= 0 {
+			// -1 if an error occurred (EBADF, EFAULT, EINVAL) or the call was
+			// interrupted by a signal (EINTR).
+			// 0 is the timeout occurred. In this case there's no timeout specified.
+			// Still handle this explicitly in case a timeout could be triggered by
+			// external events, like system sleep.
+			continue
+		}
+		if err != nil {
+			// TODO(maruel): It'd be nice to be able to surface this.
+			// This may cause a busy loop. Hopefully the user will notice and will
+			// fix their code.
+			// This can happen when removeFd() is called, in this case silently
+			// ignore the error.
+			continue
+		}
+
+		now := time.Now()
+		// Create a look up table with the lock, so that then the channel can be
+		// pushed to without the lock.
+		if cap(lookups) < n {
+			lookups = make([]lookup, 0, n)
+		} else {
+			lookups = lookups[:0]
+		}
+
+		e.mu.Lock()
+		for _, ev := range events[:n] {
+			ep := epollEvent(ev.Events)
+			// Skip over file descriptors that are not present.
+			c, ok := e.fds[ev.Fd]
+			if !ok {
+				// That's a race condition where the file descriptor was removed by
+				// removeFd() but it still triggered. Ignore this event.
+				continue
+			}
+			// Look at the event to determine if it's worth sending a pulse. It's
+			// maybe not worth it.
+			if ep&epollERR != 0 {
+				// TODO(maruel): We should stop listening to this file descriptor.
+				// Same for epollHUP.
+			}
+			// pipe and sockets trigger epollIN and epollOUT, but GPIO sysfs
+			// triggers epollPRI.
+			if ep&(epollPRI|epollIN|epollOUT) != 0 {
+				lookups = append(lookups, lookup{c: c, event: ep})
+			}
+		}
+		e.mu.Unlock()
+
+		// Once the lock is released, send the events. It's up to the client to
+		// interpret the epoll flag.
```

Comment:
Client does not seem to get event, only time, or am I missing something important?

Reply:
Thanks for catching, I used to send the epoll event but decided against now, so that explains why the comment was misleading.